### ISSUE TITLE
Vector3::Transform bug

### ISSUE THREAD

**Issue by Const-me** [mono/MonoGame#4499]
Matrix:
( -1.799367  -0.004297225  -0.003838511  -0.003838127 )  
( -0.007210343  1.072388  0.9579144  0.9578186 )  
( 0  3.574656  -0.2873766  -0.2873479 )  
( 0  0  5.210674  5.220153 )
Vector3: zero

Expected result, as produced by XMVector3TransformCoord API:
{ 0, 0, 0.998184144 }
Actual result by MonoGame code:
{ 0, 0, 5.210674 }

Fix:

	public static void Transform(ref Vector3 position, ref Matrix matrix, out Vector3 result)
	{
		Vector4 v4 = Vector4.Transform( position, matrix );
		float wi = 1.0f / v4.W;
		result.X = v4.X * wi;
		result.Y = v4.Y * wi;
		result.Z = v4.Z * wi;
	}

**Comment by tomspilman** [comment]
Interesting... can someone add this into a unit test and verify @Const-me findings against the result form XNA?

**Comment by InfiniteProductions** [comment]
Hi,

I don't think there is any issue here, at least with MonoGame itself as:
-XMVector3TransformCoord is a C++ API (from DirectXMath)
-XNA 4.0 framework itself, on windows, gives me {X:0 Y:0 Z:5,210674} which is EXACTLY the MG results @Const-me got (with MG, according to him).

It may be a XNA issue which MG may or may not fix (keep it may be best, adding a bug free method with a different name which will break compatibility). I don't know who's right or wrong here.

code used is:

           Matrix constmemtest = new Matrix( -1.799367f, -0.004297225f, -0.003838511f, 
           -0.003838127f,   -0.007210343f, 1.072388f, 0.9579144f, 0.9578186f,
           0f, 3.574656f, -0.2873766f, -0.2873479f, 0, 0, 5.210674f, 5.220153f);

            Vector3 consmevtest = Vector3.Zero;
            Vector3 constmeres;

            Vector3.Transform(ref consmevtest, ref constmemtest, out constmeres);
            Debug.Print(string.Format("constme RES={0}", constmeres));

**Comment by tomspilman** [comment]
Thanks @InfiniteProductions !

I think this is a misunderstanding of what `XMVector3TransformCoord` does verse what `Vector3.Transform` does.  From what I read `XMVector3TransformCoord` divides by resulting xyz by the w value...  XNA/MonoGame's `Vector3.Transform` does not do that.

I bet that if you test using `Vector4.Transform` passing the W as 1 you get the same results as `XMVector3TransformCoord`.

**Comment by Const-me** [comment]
Transformation matrices are 4x4.
They transform [homogeneous coordinates](https://en.wikipedia.org/wiki/Homogeneous_coordinates).

To make homogeneous coordinate from XYZ coordinate, you set W value to one.

Both XNA and D3D do that.

Both XNA and D3D then multiply homogeneous vector4 by that matrix, and produce equal result, which is { 0, 0, 5.210674, 5.220153 }

However, XMVector3TransformCoord then converts homogeneous coordinates back to XYZ. Here’s quote [from documentation](https://msdn.microsoft.com/en-us/library/windows/desktop/microsoft.directx_sdk.transformation.xmvector3transformcoord(v=vs.85).aspx): **projecting the result back into w = 1**.

XNA just drops W component, and only return XYZ values. For matrices like that, XNA returns complete garbage instead of transformed vector.

You’re saying that garbage is by design?

**Comment by tomspilman** [comment]
> XNA just drops W component, and only return XYZ values.

It doesn't do that...  it assumes W is 1.  As you can see in our implementation which returns the same results as XNA:

https://github.com/mono/MonoGame/blob/develop/MonoGame.Framework/Vector3.cs#L942

It just doesn't do the "Coord" part and divide the results by W.


> You’re saying that garbage is by design?

You are saying it is garbage by design.

I'm saying you're misunderstanding what it is and when to use it.  This same code exists in every game engine and library.

It is by design in Microsoft's new Numerics library:

https://github.com/mono/MonoGame/blob/develop/MonoGame.Framework/Vector3.cs#L942


And by design in Ogre3D:

https://bitbucket.org/sinbad/ogre/src/292c4b5c200489f429f10c89294c56b9a1ddf821/OgreMain/include/OgreMatrix4.h?at=default&fileviewer=file-view-default#OgreMatrix4.h-633


And by design in Irrlicht:

http://svn.code.sf.net/p/irrlicht/code/trunk/include/matrix4.h

I'm sure you can find the same code in Unreal Engine, Unity3D, etc.

**Comment by Const-me** [comment]
> It doesn't do that

It does.
Just call Vector4.Transform( Vector4.UnitW, thatMatrix ), and in the X,Y and Z of the result you'll see the same values as produced by  Vector3.Transform( Vector3.Zero, thatMatrix )

> it assumes W is 1

That’s correct assumption.
Incorrect assumption, the code assumes the matrix has M14=M24=M34=0, and M44=1.

> by design in Ogre3D

Technically, Ogre3D is fine: they wrote in the comments their method only works with affine matrices.
Obviously, my matrix isn’t affine.
My matrix includes perspective projection (it was in fact model * view * projection matrix).
That’s why that code produces garbage while transforming a 3D vector with it.

**Comment by tomspilman** [comment]
> Incorrect assumption, the code assumes the matrix has M14=M24=M34=0, and M44=1.

Yes... that would be incorrect to think.  The current code clearly does not do that.


> My matrix includes perspective projection (it was in fact model * view * projection matrix).

Maybe we should clear up in the docs that these are all affine transforms and not meant for general 4x4 projective transformations.

We should instead recommend people look at `Vector4.Transform(Vector3, Matrix)` or `Vector4.Transform(Vector4, Matrix)` if they want the full general transform.  That is what those are for since they return the W component results.

> That’s why that code produces garbage while transforming a 3D vector with it.

Right...  the code is working correctly if you were looking for an affine transform.  It just isn't documented correctly.

**Comment by Const-me** [comment]
> that would be incorrect to think. The current code clearly does not do that.

But it does.

Open the source code and see yourself.

You won’t see any reference to M14, M24, M34, or M44 matrix fields. That’s because currently, the code ignores the last row of the input matrix, it incorrectly assumes it to be {0,0,0,1} If it's indeed {0,0,0,1} the result is correct.

> the code is working correctly if you were looking for an affine transform. It just isn't documented correctly.

That's one way to say that.

I’d say “the current code produces complete garbage when called with a perspective matrix argument”.

**Comment by tomspilman** [comment]
> You won’t see any reference to M14, M24, M34, or M44 matrix fields. 

Right... because those would be stored in the W of the result... which doesn't exist for a Vector3.  The math is omitted because it doesn't affect the result.

> I’d say “the current code produces complete garbage when called with a perspective matrix argument”.

Then so does that function in every other game engine then.  You are right and the world is wrong! :)

**Comment by KonajuGames** [comment]
We will clarify the documentation and this issue can be considered resolved.

**Comment by Const-me** [comment]
Indeed.

Why bother fixing bugs when you can leave them in the source code, only document?

This way, another developer will spend an hour debugging the code, and then will find out it’s his own fault.

**Comment by tomspilman** [comment]
> Why bother fixing bugs

We fix bugs all the time... we love to fix bugs.  We don't see any bug here.

If you think there is a bug please show me what it is.  Here is the current implementation of the function:

```
        public static void Transform(ref Vector3 position, ref Matrix matrix, out Vector3 result)
        {
            var x = (position.X * matrix.M11) + (position.Y * matrix.M21) + (position.Z * matrix.M31) + matrix.M41;
            var y = (position.X * matrix.M12) + (position.Y * matrix.M22) + (position.Z * matrix.M32) + matrix.M42;
            var z = (position.X * matrix.M13) + (position.Y * matrix.M23) + (position.Z * matrix.M33) + matrix.M43;
            result.X = x;
            result.Y = y;
            result.Z = z;
        }
```

How would you change that to fix the bug?

**Comment by Const-me** [comment]
I’ve posted the fixed version in my original message.

If however you prefer copy-pasting code over reusing code, here’s the copy-paste friendly version for you:

```
public static void Transform(ref Vector3 position, ref Matrix matrix, out Vector3 result)
{
    float x = (position.X * matrix.M11) + (position.Y * matrix.M21) + (position.Z * matrix.M31) + matrix.M41;
	float y = (position.X * matrix.M12) + (position.Y * matrix.M22) + (position.Z * matrix.M32) + matrix.M42;
	float z = (position.X * matrix.M13) + (position.Y * matrix.M23) + (position.Z * matrix.M33) + matrix.M43;
	float w = (position.X * matrix.M14) + (position.Y * matrix.M24) + (position.Z * matrix.M34) + matrix.M44;
	float iw = 1.0f / w;
	result.X = x * iw;
    result.Y = y * iw;
    result.Z = z * iw;
}

```

**Comment by tomspilman** [comment]
> here’s the copy-paste friendly version for you 

So what you want a new function that does a general 4x4 transform with normalization on a `Vector3`.  Neither XNA or MonoGame contain such a function and it might not be a bad idea to include it with proper documentation.

Again the existing function you feel is "garbage" is still correct.  It is specifically a optimized transform for affine matrices...  something you do a lot of on the CPU.  It will not be changing other than to more clearly document it.

**Comment by Const-me** [comment]
Indeed, compared to my proposed version, your current optimized version saves six floating-point multiplications, one division, and three additions.

However, that optimization causes the function to produce incorrect result with some input data.

Don’t you think we have a classic case of premature optimization here?

**Comment by WardBenjamin** [comment]
@Const-me I'm not a project maintainer, so I can be a little more crude and direct: We're not changing this, it's as it's meant to be. This functionality is exclusively meant for affine matrices and vectors, not what you're doing. It's not a bug. *(Hint: There's Vector4.Transform(Vector3, Matrix4) for exactly what you're trying to do.)*

**THE FUNCTION IS CORRECT. END OF STORY.**

**Comment by tomspilman** [comment]
@WardBenjamin - I know you are trying to help.  Just let's not make this escalate further.

>  that optimization causes the function to produce incorrect result with some input data.

Any function produces incorrect results if you are expecting it to do something it is not meant to do.  You are expecting it to return the Vector3 in the W normalized form... it does not do that as it assumes the input matrix is affine.

> Don’t you think we have a classic case of premature optimization here?

Not at all.  Again this same function exists in every game engine for a reason.  Doing useless math is always unnecessary.

**Comment by WardBenjamin** [comment]
@tomspilman fair enough!

**Comment by HyunMi027** [comment]
As other have pointed out already `TransformCoord` is not the same as `Transform`. `TransformCoord` considers the **translation** that is "stored" in the w component. `Transform` on the other hand does not do this. 

Let us consider a `Vector3` which I'll call **v** and a Matrix which I'll call **m** and the output which I'll call **o** now let us put this through both of the functions ( braces are just for clarity sake and bold highlights are to denote important differences between the 2).

**TransformCoord**
***
o<sub>x</sub> = (v<sub>x</sub> * m<sub>11</sub>) + (v<sub>y</sub> * m<sub>21</sub>) + (v<sub>z</sub> * m<sub>31</sub>) **+ (m<sub>41</sub>)**
o<sub>y</sub> = (v<sub>x</sub> * m<sub>12</sub>) + (v<sub>y</sub> * m<sub>22</sub>) + (v<sub>z</sub> * m<sub>32</sub>) **+ (m<sub>42</sub>)**
o<sub>z</sub> = (v<sub>x</sub> * m<sub>13</sub>) + (v<sub>y</sub> * m<sub>23</sub>) + (v<sub>z</sub> * m<sub>33</sub>) **+ (m<sub>43</sub>)**

**Transform**
***
o<sub>x</sub> = (v<sub>x</sub> * m<sub>11</sub>) + (v<sub>y</sub> * m<sub>21</sub>) + (v<sub>z</sub> * m<sub>31</sub>)
o<sub>y</sub> = (v<sub>x</sub> * m<sub>12</sub>) + (v<sub>y</sub> * m<sub>22</sub>) + (v<sub>z</sub> * m<sub>32</sub>)
o<sub>z</sub> = (v<sub>x</sub> * m<sub>13</sub>) + (v<sub>y</sub> * m<sub>23</sub>) + (v<sub>z</sub> * m<sub>33</sub>)

The key difference is the extra term of **+ (m<sub>4</sub>..** . One might ask the question why is there no multiplication happening there? Well there is, but you are multiplying by 1 ( as others pointed out as well ). So we can conclude that `TransformCoord` is not the same as `Transform`.

Hope this clarifies as to why they are not the same.
***

**tl;dr:** `TransformCoord` **!=** `Transform`

**Comment by Const-me** [comment]
@HyunMi027 in MG, the functions are named differently.
The function you’ve listed as TransformCoord is named Vector3.Transform.
The function you’ve listed as Transform is named Vector3.TransformNormal.

Second, the bug isn’t relevant to the difference between Transform and TransformNormal.

The bug is, when you call Vector3.Transform  with a matrix where 
m <sub>14</sub> != 0 || m<sub>24</sub> != 0 || m<sub>34</sub> != 0 || m<sub>44</sub> != 1
then Vector3.Transform will return garbage instead of the result, because it doesn’t project the result back into W = 1, but instead just discards W.

**Comment by nkast** [comment]
@Const-me,
 Why compare `Vector3.Transform()` with `::XMVector3TransformCoord()` in the first place since DirectXMath has `::XMVector3Transform()`?

https://msdn.microsoft.com/en-us/library/windows/desktop/microsoft.directx_sdk.transformation.xmvector3transform(v=vs.85).aspx
Transforms a 3D vector by a matrix.
XMVector3Transform ignores the w component of the input vector, and uses a value of 1 instead. The w component of the returned vector may be non-homogeneous (!= 1.0).

https://msdn.microsoft.com/en-us/library/windows/desktop/microsoft.directx_sdk.transformation.xmvector3transformcoord(v=vs.85).aspx
Transforms a 3D vector by a given matrix, projecting the result back into w = 1
XMVector3TransformCoord ignores the w component of the input vector, and uses a value of 1.0 instead. The w component of the returned vector will always be 1.0.

From what I get `Vector3.Transform()` is similar to `::XMVector3Transform()`. MG doesn't have a `Vector3.TransformCord()`.

As to whether or not `Transform(v3, m4x4, out v3)` return garbage isn't something we can decide upon. It's nothing more than a 'Helper method', math doesn't define a multiplication between Vector3 and Matrix4x4.

**Comment by HyunMi027** [comment]
@Const-me 

> then Vector3.Transform will return garbage instead of the result, because it doesn’t project the result back into W = 1, but instead just discards W.
Well that was the clue of the story, it discards w on purpose not by a design flaw or bug.

**Comment by tgjones** [comment]
So... the conclusion here is that XNA and MonoGame don't have have an equivalent to `XMVector3TransformCoord`, but someone could, if they wanted to, add a `Vector3.TransformCoord` to MonoGame?

Personally I don't see the need for it - it's fairly rare that you do anything on the CPU that results in `.W` being anything other than 1, and if you do need to do it, it's trivial to create your own helper method.

In any case, I think if this issue is left open, it should be renamed to indicate that it's a feature request for a new method, not a bug in an existing one.

**Comment by Const-me** [comment]
> fairly rare that you do anything on the CPU that results in  .W  being anything other than 1

If you count calls, sure, it's rare.
If you count projects, no it's not.
Many projects need mouse picking, or other screen-related transforms/manipulations.
That's usually done on CPU.
Last row is not 1 in model * view * projection matrix if the projection's perspective.

Sure it's easy to fix - but only if you know the exact reason.

If for example debug build of MG would check the input matrix, and throw the exception if the last row isn't 1 - that would also solve this problem without sacrificing the runtime performance.

**Comment by KonajuGames** [comment]
This thread will end now.  We know where everyone stands and what needs to be done.

- Vector3.Transform will remain unchanged.  It's behaviour is intended and correct for its purpose.  The documentation for Vector3.Transform will be updated to include clarification that it ignores the w component.
- Add Vector3.TransformCoord that does the same as XMVector3TransformCoord by mapping the result back to w = 1.  Documentation to be clear about its behaviour.

Separate issues will be raised for these items.  This thread is now closed.
